tracing-appender/src/rolling/builder.rs

```diff
@@ -0,0 +1,147 @@
+//! Builder struct for `RollingFileAppender`
+//!
+//! Gives access to setting additional options which are not avaible using standard interface.
+//! Currently it is the only way to enable compression of logs.
+use crate::rolling::{create_writer_file, Inner, RollingFileAppender, Rotation};
+use crate::sync::RwLock;
+use std::path::Path;
+use std::sync::atomic::AtomicUsize;
+use time::OffsetDateTime;
+
+#[cfg(feature = "compression_gzip")]
+use {
+    crate::rolling::compression::CompressionConfig, crate::rolling::compression::CompressionOption,
+};
+
+use crate::writer::WriterChannel;
+
+/// Struct for keeping temporary values of `RollingFileAppender`.
+///
+/// Note that `log_directory` and `log_filename_prefix` are obligatory parameters and should
+/// be passed into the constructor of `RollingFileAppenderBuilder`.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct RollingFileAppenderBuilder {
+    log_directory: String,
+    log_filename_prefix: String,
+    rotation: Option<Rotation>,
+    #[cfg(feature = "compression_gzip")]
+    compression: Option<CompressionConfig>,
+}
+
+impl RollingFileAppenderBuilder {
+    /// Creates a new `RollingFileAppnderBuilder`
+    ///
+    /// It was introduced to open up the possibility to use `compression` without
+    /// breaking the current interface.
+    ///
+    /// Note that `compression` module is enabled by using an optional feature flag
+    /// `compression_gzip` (for gzip algorithm)
+    ///
+    /// # Examples
+    /// ```rust
+    /// use tracing_appender::builder::RollingFileAppenderBuilder;
+    /// use tracing_appender::compression::CompressionOption;
+    /// use tracing_appender::rolling::Rotation;
+    /// let builder = RollingFileAppenderBuilder::new("/var/tmp", "my-app")
+    ///     .rotation(Rotation::DAILY)
+    ///     .compression(CompressionOption::GzipFast);
+    /// ```
```

Comment:
How about something like this:
```suggestion
    /// Returns a new builder for configuring a [`RollingFileAppender`], with
    /// the provided log directory and filename prefix.
    ///
    /// Calling [`build`] on the returned builder will construct a new
    /// [`RollingFileAppender`] that writes to files in the provided directory.
    /// By default, log files will never be rotated and compression will not be
    /// enabled. A rotation policy can be added to the builder using the
    /// [`rotation`] method. When the "compression" feature flag is enabled,
    /// compression can be configured using the [`compression`] method.
    ///
    /// # Panics
    ///
    /// This function panics if the provided log directory or log file prefix
    /// are not valid UTF-8.
    ///
    /// # Examples
    ///
    /// Building a `RollingFileAppender` with the default configuration:
    ///
    /// ```rust
    /// use tracing_appender::rolling::RollingFileAppenderBuilder;
    /// let appender = RollingFileAppenderBuilder::new("/var/tmp", "my-app")
    ///     .build();
    /// ```
    ///
    /// Enabling compression:
    ///
    /// ```rust
    /// use tracing_appender::{
    ///     rolling::{RollingFileAppenderBuilder, Rotation},
    ///     compression::CompressionOption,
    /// };
    /// let appender = RollingFileAppenderBuilder::new("/var/tmp", "my-app")
    ///     .rotation(Rotation::DAILY)
    ///     .compression(CompressionOption::GzipFast)
    ///     .build();
    /// ```
```